### PR TITLE
Use real samples on QPU device

### PR DESCRIPTION
**Pull request by dabacon**
In the case of the actual QPU, instead of resample from the returned results, one should just use the actual samples.  

This is a bit convoluted because the api currently returns probabilities instead of counts, so should be upgraded if/when counts are exposed.

### CODE REVIEW

**Review by josh146 — COMMENTED** [review]
Thanks @dabacon! If I recall, this is something we debated initially while putting the plugin together, as we weren't sure how to extract samples from the API.

Quick question; from looking at the change, the new logic is that it uses the counts to 'repeat' the measured computational basis states in order from 0 to N-1, and then converts them into the equivalent binary representation.

Would the inherent ordering of the returned samples be an issue? I can't formulate a strong reason _why_ this would be the case, but was just curious if it makes sense to do an in-place [`np.random.shuffle`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.shuffle.html) before returning them.

**Comment by dabacon** [comment]
Some apis do return the results where the order is the order the experiments were run, but IonQ's api doesn't return this order information. For such an API you can analyze things for drift over time.  Fun stuff :) 

The one reason I can think about for doing a shuffle is that this can handle bugs where someone slices the results expected that to be equivalent to asking for a sub-sample.  The downside is that in this case it introduces a dependency on the numpy's random number generator, which is a bit strange.  I guess one solution is that we could set the seed explicitly for this.  But I guess the question is whether you think there will be code that does this slicing.  If so I think the random shuffle is the right way to go.

**Comment by josh146** [comment]
> The one reason I can think about for doing a shuffle is that this can handle bugs where someone slices the results expected that to be equivalent to asking for a sub-sample. 

Good point, I think this is an important enough reason to do the shuffling. In fact, there are cases where I realize now this _could be required_. One example is any case where you are distributing expectation values across sub-batches of shots from a single job --- e.g., the extreme case of single-shot gradient variance is required to do shot-adaptive optimization.

File: pennylane_ionq/device.py [hunk 1]
```diff
@@ -246,3 +247,18 @@ class QPUDevice(IonQDevice):
 
     def __init__(self, wires, *, shots=1024, api_key=None):
         super().__init__(wires=wires, target="qpu", shots=shots, api_key=api_key)
+
+    def generate_samples(self):
+        """Generates samples from the qpu.
+
+        Note that the order of the samples returned here is not indicative of the order in which
+        the experiments were done, but is instead controlled by a random shuffle (and hence
+        set by numpy random seed.)
```
Comment (co9olguy) [inline pennylane_ionq/device.py:256]:
```suggestion
        set by numpy random seed).
```

**Review by co9olguy — COMMENTED** [review]
Thanks @dabacon! Looks good. 

We should add a (mocked) test of this newly written method to `test_device.py` before merging

**Comment by dabacon** [comment]
Shuffle is in place!  Fixed this and the wrong period placement.

**Review by co9olguy — APPROVED** [review]

**Comment by co9olguy** [comment]
Thanks @dabacon!

**Pull request merged**